File: test/ModularPipelines.UnitTests/Dependencies/DependsOnModulesWithTagAttributeTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

/// <summary>
/// Unit tests for <see cref="DependsOnModulesWithTagAttribute"/>.
/// </summary>
public class DependsOnModulesWithTagAttributeTests : TestBase
{
    [Test]
    public async Task ShouldDependOn_ModuleHasTag_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithTagAttribute("database");
        var context = new MockDependencyContext().WithTags(typeof(TestModule), "database");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleLacksTag_ReturnsFalse()
    {
        var attr = new DependsOnModulesWithTagAttribute("database");
        var context = new MockDependencyContext().WithTags(typeof(TestModule), "other");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasNoTags_ReturnsFalse()
    {
        var attr = new DependsOnModulesWithTagAttribute("database");
        var context = new MockDependencyContext();

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsFalse();
    }

    [Test]
    public async Task ShouldDependOn_CaseInsensitive_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithTagAttribute("DATABASE");
        var context = new MockDependencyContext().WithTags(typeof(TestModule), "database");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task ShouldDependOn_ModuleHasMultipleTagsIncludingMatch_ReturnsTrue()
    {
        var attr = new DependsOnModulesWithTagAttribute("database");
        var context = new MockDependencyContext().WithTags(typeof(TestModule), "infrastructure", "database", "critical");

        var result = attr.ShouldDependOn(typeof(TestModule), context);

        await Assert.That(result).IsTrue();
    }

    [Test]
    public async Task Constructor_WithNullTag_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesWithTagAttribute(null!))
            .ThrowsException();
    }

    [Test]
    public async Task Constructor_WithEmptyTag_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesWithTagAttribute(string.Empty))
            .ThrowsException();
    }

    [Test]
    public async Task Constructor_WithWhitespaceTag_ThrowsArgumentException()
    {
        await Assert.That(() => new DependsOnModulesWithTagAttribute("   "))
            .ThrowsException();
    }

    [Test]
    public async Task Tag_Property_ReturnsConstructorValue()
    {
        var attr = new DependsOnModulesWithTagAttribute("my-tag");

        await Assert.That(attr.Tag).IsEqualTo("my-tag");
    }

    private class TestModule { }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/DependsOnModulesWithTagAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open System
open ModularPipelines.Attributes
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

type TagAttributeTestModule() =
    class
    end

type DependsOnModulesWithTagAttributeTests() =
    [<Test>]
    member _.ShouldDependOn_ModuleHasTag_ReturnsTrue() = async {
        let attr = DependsOnModulesWithTagAttribute("database")
        let context = MockDependencyContext().WithTags(typeof<TagAttributeTestModule>, "database")
        let result = attr.ShouldDependOn(typeof<TagAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleLacksTag_ReturnsFalse() = async {
        let attr = DependsOnModulesWithTagAttribute("database")
        let context = MockDependencyContext().WithTags(typeof<TagAttributeTestModule>, "other")
        let result = attr.ShouldDependOn(typeof<TagAttributeTestModule>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasNoTags_ReturnsFalse() = async {
        let attr = DependsOnModulesWithTagAttribute("database")
        let context = MockDependencyContext()
        let result = attr.ShouldDependOn(typeof<TagAttributeTestModule>, context)
        do! check(Assert.That(result).IsFalse())
    }

    [<Test>]
    member _.ShouldDependOn_CaseInsensitive_ReturnsTrue() = async {
        let attr = DependsOnModulesWithTagAttribute("DATABASE")
        let context = MockDependencyContext().WithTags(typeof<TagAttributeTestModule>, "database")
        let result = attr.ShouldDependOn(typeof<TagAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.ShouldDependOn_ModuleHasMultipleTagsIncludingMatch_ReturnsTrue() = async {
        let attr = DependsOnModulesWithTagAttribute("database")
        let context = MockDependencyContext().WithTags(typeof<TagAttributeTestModule>, "infrastructure", "database", "critical")
        let result = attr.ShouldDependOn(typeof<TagAttributeTestModule>, context)
        do! check(Assert.That(result).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithNullTag_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesWithTagAttribute(null) |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithEmptyTag_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesWithTagAttribute(String.Empty) |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Constructor_WithWhitespaceTag_ThrowsArgumentException() = async {
        let mutable threw = false

        try
            DependsOnModulesWithTagAttribute("   ") |> ignore
        with :? ArgumentException ->
            threw <- true

        do! check(Assert.That(threw).IsTrue())
    }

    [<Test>]
    member _.Tag_Property_ReturnsConstructorValue() = async {
        let attr = DependsOnModulesWithTagAttribute("my-tag")
        do! check(StringEqualsAssertionExtensions.IsEqualTo(Assert.That(attr.Tag), "my-tag"))
    }
